clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java

```diff
@@ -0,0 +1,263 @@
+ * <em>{@link Assignment#topicPartitions() target assignment}</em>.
+ *
+ * <p/>
+ *
+ * Internally, reconciliation requires the following steps:
+ *
+ * <ol>
+ *     <li>
+ *         On the background thread, upon receipt of a new assignment from the group coordinator, the
+ *         {@link MembershipManager} should call {@link #startReconcile(Assignment)} to start reconciliation.
+ *     </li>
+ *     <li>
+ *         Internally, the partitions to revoke are determined via {@link #getPartitionsToRevoke(Assignment)};
+ *         these are the partitions in the current assignment that are <em>not in</em> the target assignment.
+ *         Next, we calculate the partitions to assign using {@link #getPartitionsToAssign(Assignment)};
+ *         these are the partitions in the target assignment that are <em>not in</em> the current assignment.
+ *     </li>
+ *     <li>
+ *         Send a {@link PartitionReconciliationStartedEvent} so that the application thread will execute the
+ *         {@link ConsumerRebalanceListener} callback methods.
+ *     </li>
+ *     <li>
+ *         On the application thread, when the {@link PartitionReconciliationStartedEvent} is received, execute the
+ *         {@link ConsumerRebalanceListener#onPartitionsRevoked(Collection)} and
+ *         {@link ConsumerRebalanceListener#onPartitionsAssigned(Collection)} callback methods
+ *     </li>
+ *     <li>
+ *         Enqueue a corresponding {@link PartitionReconciliationCompleteEvent} so that the background thread
+ *         will know the listener was invoked and the result of the invocation
+ *     </li>
+ *     <li>
+ *         On the background thread, process the {@link PartitionReconciliationCompleteEvent}, which should call the
+ *         {@link MembershipManager#completeReconcile(Set, Set, Optional)} method. This method will call
+ *         {@link #completeReconcile(Set, Set)} to remove the revoked partitions and add the assigned partitions in the
+ *         {@link SubscriptionState#assignFromSubscribed(Collection) current assignment} and then make a note to
+ *         send a {@link ConsumerGroupHeartbeatRequest} to the group coordinator on its next pass of
+ *         {@link HeartbeatRequestManager#poll(long)}
+ *     </li>
+ * </ol>
+ *
+ * <p/>
+ *
+ * The comparison against the {@link SubscriptionState#assignedPartitions() current set of assigned partitions} and
+ * the {@link Assignment#topicPartitions() target set of assigned partitions} is performed by essentially
+ * <em>flattening</em> the respective entries into two sets of {@link TopicPartition partitions}
+ * which are then compared using basic {@link Set} comparisons.
+ */
+public class AssignmentReconciler {
+
+    private final Logger log;
+    private final SubscriptionState subscriptions;
+    private final ConsumerMetadata metadata;
+    private final BlockingQueue<BackgroundEvent> backgroundEventQueue;
```

Comment:
We have a class called errorEventHandler - maybe we should rename/refacotr it and use it here.

Reply:
It's been changed a bit in the forthcoming fetch request manager PR to be more generic sense we need to also pass back non-error-related events to the application thread.